### PR TITLE
Improve show peer ids feature

### PR DESCRIPTION
**Pull request by ENCRYPTEDFOREVER**
This PR changes appearance of "Show Peer IDs" experimental feature, adding it's own row in UI.
This makes the feature look more natural and consistent with the app, also enabling copy on click

### CODE REVIEW

File: app/src/main/java/org/thunderdog/challegram/ui/SettingsController.java [hunk 1]
```diff
@@ -877,6 +888,7 @@ public void onMyUserBioUpdated (@Nullable TdApi.FormattedText newBio) {
     setBio(newBio);
   }
 
+  private @Nullable Long myId;
```
Comment (vkryl) [inline app/src/main/java/org/thunderdog/challegram/ui/SettingsController.java:891]: No need in field

File: app/src/main/java/org/thunderdog/challegram/ui/SettingsController.java [hunk 2]
```diff
@@ -885,6 +897,13 @@ private void initMyUser () {
     TdApi.User user = tdlib.myUser();
     setUsername(user);
     setPhoneNumber(user);
+    setUserId(user);
+  }
+
+  private boolean setUserId(@Nullable TdApi.User myUser) {
+    Long oldId = myId;
+    myId = myUser != null ? myUser.id : null;
+    return !Objects.equals(oldId, myId);
```
Comment (vkryl) [inline app/src/main/java/org/thunderdog/challegram/ui/SettingsController.java:906]: Rely on `tdlib.myUserId(true)` when needed

File: app/src/main/java/org/thunderdog/challegram/ui/SettingsController.java [hunk 3]
```diff
@@ -862,6 +870,9 @@ public void onMyUserUpdated (final TdApi.User myUser) {
     }
     runOnUiThreadOptional(() -> {
       updateHeader();
+      if (setUserId(myUser)) {
+        adapter.updateValuedSettingById(R.id.btn_peer_id);
+      }
```
Comment (vkryl) [inline app/src/main/java/org/thunderdog/challegram/ui/SettingsController.java:875]: Can't change during runtime (except being logged out, where it gets replaced to 0, in which case there's no need to update UI)

File: app/src/main/java/org/thunderdog/challegram/ui/ProfileController.java [hunk 4]
```diff
@@ -765,6 +765,34 @@ private String getProfileUsername () {
     return tdlib.chatUsername(chat.id);
   }
 
+  private long getPeerId () {
+    return user != null ? user.id : secretChat != null ? secretChat.id : chat != null ? chat.id : 0;
```
Comment (vkryl) [inline app/src/main/java/org/thunderdog/challegram/ui/ProfileController.java:769]: `chat.id` isn't the same as `basicGroup.id` and `supergroup.id`, and shouldn't be returned by this method / viewable as peer ID.

**Review by vkryl — CHANGES_REQUESTED** [review]
I didn't review or test the UI, however, there are clear issues in the code (pointed out in separate comments).

File: app/src/main/java/org/thunderdog/challegram/ui/ProfileController.java (hunk 4 above)
Comment (ENCRYPTEDFOREVER) [reply app/src/main/java/org/thunderdog/challegram/ui/ProfileController.java:769]: Can you clarify?

Comment (vkryl) [reply app/src/main/java/org/thunderdog/challegram/ui/ProfileController.java:769]: Yes.

File: app/src/main/java/org/thunderdog/challegram/ui/SettingsController.java [hunk 5]
```diff
@@ -974,6 +978,11 @@ public void onClick (View v) {
       EditBioController c = new EditBioController(context, tdlib);
       c.setArguments(new EditBioController.Arguments(about != null ? about.text : "", 0));
       navigateTo(c);
+    } else if (viewId == R.id.btn_peer_id) {
+      long myId = tdlib.myUserId(true);
+      if (myId!= 0) {
```
Comment (vkryl) [inline app/src/main/java/org/thunderdog/challegram/ui/SettingsController.java:983]:
```suggestion
      if (myId != 0) {
```

File: app/src/main/java/org/thunderdog/challegram/ui/SettingsController.java [hunk 6]
```diff
@@ -81,6 +81,7 @@
 
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
```
Comment (vkryl) [inline app/src/main/java/org/thunderdog/challegram/ui/SettingsController.java:84]:
```suggestion
```

File: app/src/main/java/org/thunderdog/challegram/ui/ProfileController.java [hunk 7]
```diff
@@ -766,28 +766,36 @@ private String getProfileUsername () {
...
+    } else if (supergroup != null) {
+      return supergroup.id;
+    } else if (user != null) {
+      return user.id;
+    } else {
+      return 0;
+    }
   }
 
   private int getPeerTypeStringResourceId () {
```
Comment (ENCRYPTEDFOREVER) [inline app/src/main/java/org/thunderdog/challegram/ui/ProfileController.java:782]: Btw, aren't there also megagroups and gigagroups? Should I include them

**Comment by vkryl** [comment]
To speed up the review, please include the screenshots of the changes.

**Comment by ENCRYPTEDFOREVER** [comment]
> To speed up the review, please include the screenshots of the changes (how profile and settings screens look after changes).

**Comment by vkryl** [comment]
Looks okay to me. @dotvhs aye or nay?

**Comment by dotvhs** [comment]
> Looks okay to me. @dotvhs aye or nay?

Looks totally fine and also solves minor things like bio is currently being copied with userid as well as translated too.

**Comment by vkryl** [comment]
There's an issue that peer ID is shown below/above "description" randomly.  It should always be the first item below avatar, and separator between the peer ID and the next item should be shown properly. Also indexes of updated items should be updated properly. Refer to `addTopItem` and its usages as well (`index` should be incremented by `1`, if peer ID is going to be the topmost item).

**Review by vkryl — CHANGES_REQUESTED** [review]

**Comment by ENCRYPTEDFOREVER** [comment]
> ..And separator between the peer ID and the next item should be shown properly.
Like [here](https://user-images.githubusercontent.com/48645524/283197187-6cabfa44-41f0-4215-8904-0cec4a884409.png) between description and a peer id?

**Comment by vkryl** [comment]
> > ..And separator between the peer ID and the next item should be shown properly.
> 
> Like [here](https://user-images.githubusercontent.com/48645524/283197187-6cabfa44-41f0-4215-8904-0cec4a884409.png) between description and a peer id?

No. [Here](https://github.com/TGX-Android/Telegram-X/assets/6242627/8df2038f-81f5-4e2a-98b7-8ea6e47f2116) it is shown properly (the black line). On some screenshots it isn't

**Comment by ENCRYPTEDFOREVER** [comment]
> No. [Here](https://github.com/TGX-Android/Telegram-X/assets/6242627/8df2038f-81f5-4e2a-98b7-8ea6e47f2116) it is shown properly (the black line). On some screenshots it isn't

Oooh, that thing. They're so thin that I didn't event notice any on the level of brightness I was using

File: app/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -4872,8 +4874,17 @@
...
+  <string name="Experiment_PeerIdsInfo">Show peer IDs from API in profile view</string>
   <string name="Experiment_ChatFolders">Enable Chat Folders</string>
   <string name="Experiment_ChatFoldersInfo">Enable early access to folders. Appearance of this feature is planned to be changed in future versions of Telegram X, and it includes some unfinished work. Credit: [PR #297](https://github.com/TGX-Android/Telegram-X/pull/297) by nikita-toropov</string>
 
+  <string name="PeerId">Peer ID</string>
+  <string name="UserId">User ID</string>
+  <string name="SecretChatId">Secret Chat ID</string>
+  <string name="BotId">Bot ID</string>
+  <string name="ChatId">Chat ID</string>
+  <string name="BasicGroupId">Basic Group ID</string>
```
Comment (vkryl) [inline app/src/main/res/values/strings.xml:4886]:
```suggestion
  <string name="BasicGroupId">Group ID</string>
```

File: app/src/main/res/values/strings.xml [hunk 9]
```diff
@@ -4872,8 +4874,17 @@
   <string name="ExperimentalSettings">Experimental settings</string>
   <string name="ExperimentalSettingsWarning">**Warning:** those are experimental settings. Some may not work. Others may break the app. Any of them might disappear without a trace.\n\nUse at your own risk.</string>
   <string name="Experiment_PeerIds">Show Peer IDs in Profiles</string>
-  <string name="Experiment_PeerIdsInfo">Show peer IDs from API below their Bio / Description</string>
+  <string name="Experiment_PeerIdsInfo">Show peer IDs from API in profile view</string>
```
Comment (vkryl) [inline app/src/main/res/values/strings.xml:4877]:
```suggestion
  <string name="Experiment_PeerIdsInfo">Show API identifiers of users and chats in their profiles</string>
```

File: app/src/main/res/values/strings.xml [hunk 10]
```diff
@@ -4870,8 +4872,17 @@
...
-  <string name="Experiment_PeerIdsInfo">Show peer IDs from API below their Bio / Description</string>
+  <string name="Experiment_PeerIdsInfo">Show API identifiers of users and chats in their profiles</string>
   <string name="Experiment_ChatFolders">Enable Chat Folders</string>
   <string name="Experiment_ChatFoldersInfo">Enable early access to folders. Appearance of this feature is planned to be changed in future versions of Telegram X, and it includes some unfinished work. Credit: [PR #297](https://github.com/TGX-Android/Telegram-X/pull/297) by nikita-toropov</string>
 
+  <string name="PeerId">Peer ID</string>
+  <string name="UserId">User ID</string>
+  <string name="SecretChatId">Secret Chat ID</string>
+  <string name="BotId">Bot ID</string>
+  <string name="ChatId">Chat ID</string>
```
Comment (vkryl) [inline app/src/main/res/values/strings.xml:4883]:
```suggestion
```

File: app/src/main/java/org/thunderdog/challegram/ui/ProfileController.java [hunk 11]
```diff
@@ -2294,6 +2330,10 @@ private ListItem newDescriptionItem () {
     return new ListItem(ListItem.TYPE_INFO_MULTILINE, R.id.btn_description, R.drawable.baseline_info_24, isUserMode() && !TD.isBot(user) ? R.string.UserBio : R.string.Description);
   }
 
+  private ListItem newPeerIdItem () {
+    return new ListItem(ListItem.TYPE_INFO_SETTING, R.id.btn_peer_id, R.drawable.baseline_code_24, R.string.PeerId);
```
Comment (vkryl) [inline app/src/main/java/org/thunderdog/challegram/ui/ProfileController.java:2334]:
```suggestion
    return new ListItem(ListItem.TYPE_INFO_SETTING, R.id.btn_peer_id, R.drawable.baseline_identifier_24, R.string.PeerId);
```

**Review by vkryl — APPROVED** [review]
I've pushed some improvements (pretty display, optimization when setting is disabled) and bugfixes for several more issues (not all `addTopItem` calls were updated). Brief testing on the real device didn't reveal any issues.

**Pull request merged**